## Описание исследования

* Преамбула: Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать. Наша компания решила быть на волне и купила крупный сервис для чтения книг по подписке.

* Цель: Сформулировать ценностное предложение для нового продукта - сервиса для чтения книг по подписке.

* Задача: проанализировать полученную базу данных с информацией о книгах, издательствах и авторах, а также с пользовательскими обзорами книг.


**Описание данных**

Таблица `books` - содержит данные о книгах:
* `book_id` — идентификатор книги;
* `author_id` — идентификатор автора;
* `title` — название книги;
* `num_pages` — количество страниц;
* `publication_date` — дата публикации книги;
* `publisher_id` — идентификатор издателя.

Таблица `authors` - содержит данные об авторах:
* `author_id` — идентификатор автора;
* `author` — имя автора.

Таблица `publishers` - содержит данные об издательствах:
* `SQL2publisher_id` — идентификатор издательства;
* `publisher` — название издательства;

Таблица `ratings` - содержит данные о пользовательских оценках книг:
* `rating_id` — идентификатор оценки;
* `book_id` — идентификатор книги;
* `username` — имя пользователя, оставившего оценку;
* `rating` — оценка книги.

Таблица `reviews` - содержит данные о пользовательских обзорах на книги:
* `review_id` — идентификатор обзора;
* `book_id` — идентификатор книги;
* `username` — имя пользователя, написавшего обзор;
* `text` — текст обзора.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

# устанавливаем параметры
db_config = {'user': 'praktikum_student',  # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp',  # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,  # порт подключения
             'db': 'data-analyst-final-project-db'}  # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode': 'require'})

In [2]:
# Выведем первые строки из каждой таблицы

for table in ['books', 'authors', 'publishers', 'ratings', 'reviews']:
    print("")
    query = f"""SELECT * FROM {table} LIMIT 10"""
    print(f"Обзор таблицы {table}")
    display(pd.io.sql.read_sql(query, con=engine))
    print('=' * 15)


Обзор таблицы books


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116


Обзор таблицы authors


author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts


Обзор таблицы publishers


publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers


Обзор таблицы ratings


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4


Обзор таблицы reviews


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

In [3]:
# Посчитайте, сколько книг вышло после 1 января 2000 года;

query = """SELECT COUNT(*) AS books_since_2000 FROM books WHERE publication_date > '2000-01-01'"""
display(pd.io.sql.read_sql(query, con=engine))

books_since_2000
0               819

Всего в базе 819 книг, вышедших **после** 1 января 2000 года. Довольно скромный показатель.

In [4]:
# Для каждой книги посчитайте количество обзоров и среднюю оценку;

query = """SELECT title, COUNT(review_id) AS reviews, AVG(rating) AS avg_rating
FROM books
JOIN reviews USING (book_id)
JOIN ratings USING (book_id)
GROUP BY title
ORDER BY avg_rating DESC, title"""
display(pd.io.sql.read_sql(query, con=engine))

title  reviews  avg_rating
0                        A Dirty Job (Grim Reaper  #1)       16        5.00
1                A Fistful of Charms (The Hollows  #4)        4        5.00
2            A Woman of Substance (Emma Harte Saga #1)        4        5.00
3                      Act of Treason (Mitch Rapp  #9)        4        5.00
4                                        Alas  Babylon        4        5.00
..                                                 ...      ...         ...
988  The World Is Flat: A Brief History of the Twen...       12        2.25
989                                      Drowning Ruth        9        2.00
990                  His Excellency: George Washington        4        2.00
991                                              Junky        4        2.00
992                               Harvesting the Heart        4        1.50

[993 rows x 3 columns]

Книг слишком много для того, чтобы оценить всю таблицу одним взглядом, но по первым и последним строкам можно сделать вывод: средний рейтинг варьируется от 1.5 до 5.0, а также можно предположить, что усреднённый в свою очередь, он будет ближе к 5 баллам, чем к 1.5.

In [5]:
# Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

query = """WITH a AS (
SELECT publisher, COUNT(book_id) AS published_books
FROM publishers
JOIN books USING (publisher_id)
WHERE num_pages > 50
GROUP BY 1)
SELECT *
FROM a
WHERE published_books = (SELECT MAX(published_books) FROM a)"""
display(pd.io.sql.read_sql(query, con=engine))

publisher  published_books
0  Penguin Books               42

Где-то в Готэме плачет Бэтмен. Где-то икнул подниматель пингвинов. Где-то выходит из душа Пен-Пен. Где-то мистер Поппер не досчитался пингвина. Где-то ухмыльнулся 1 (прописью: один) Линус Торвальд.

Издательство Пингвин лидирует по количеству изданных "толстых" книг с показателем, который по чистой случайности равен ответу на главный вопрос жизни, вселенной и вообще.

In [6]:
# Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

query = """WITH a AS (
    SELECT book_id, COUNT(rating_id) AS n_ratings
    FROM books
    JOIN ratings USING (book_id)
    GROUP BY 1),
b AS (
    SELECT author, AVG(rating) AS rating
    FROM authors
    JOIN books USING (author_id)
    JOIN ratings USING (book_id)
    WHERE book_id IN (SELECT book_id FROM a WHERE n_ratings >= 50)
    GROUP BY author)
SELECT *
FROM b
WHERE rating = (SELECT MAX(rating) FROM b)"""
display(pd.io.sql.read_sql(query, con=engine))

author    rating
0  J.K. Rowling/Mary GrandPré  4.287097

Все любят Гарри Поттера.
Особенно Джоан Роулинг, занявшая первую строчку в списке авторов с самыми высокооценёнными книгами.

In [7]:
# Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

query = """WITH a AS (
    SELECT username, COUNT(review_id) AS n_reviews
    FROM reviews
    WHERE username IN (
        SELECT username
        FROM ratings
        GROUP BY username
        HAVING COUNT(rating_id) > 50)
    GROUP BY username)
SELECT AVG(n_reviews) AS avg_reviews
FROM a
"""
display(pd.io.sql.read_sql(query, con=engine))

avg_reviews
0    24.333333

Мы не знаем, что пишут такие активные пользователи в обзорах, но мы точно знаем, что писать обзоры сложнее, чем просто выставлять оценки: среди пользователей, которые поставили больше 50 оценок, среднее количество обзоров - ~24.3.